cmd/thanos/rule.go

```diff
@@ -318,25 +333,63 @@ func runRule(
 		// Discover and resolve query addresses.
 		addDiscoveryGroups(g, queryClient, conf.query.dnsSDInterval)
 	}
+	var (
+		appendable storage.Appendable
+		queryable  storage.Queryable
+		db         *tsdb.DB
+	)
 
-	db, err := tsdb.Open(conf.dataDir, log.With(logger, "component", "tsdb"), reg, tsdbOpts, nil)
+	rwCfgYAML, err := conf.rwConfig.Content()
 	if err != nil {
-		return errors.Wrap(err, "open TSDB")
+		return err
 	}
 
-	level.Debug(logger).Log("msg", "removing storage lock file if any")
-	if err := removeLockfileIfAny(logger, conf.dataDir); err != nil {
-		return errors.Wrap(err, "remove storage lock files")
-	}
+	if len(rwCfgYAML) > 0 {
+		var rwCfg config.RemoteWriteConfig
+		if err := yaml.Unmarshal(rwCfgYAML, &rwCfg); err != nil {
+			return err
+		}
+		walDir := filepath.Join(conf.dataDir, rwCfg.Name)
+		// flushDeadline is set to 1m, but it is for metadata watcher only so not used here.
+		remoteStore := remote.NewStorage(logger, reg, func() (int64, error) {
+			return 0, nil
+		}, walDir, 1*time.Minute, nil)
+		if err := remoteStore.ApplyConfig(&config.Config{
+			GlobalConfig:       config.DefaultGlobalConfig,
+			RemoteWriteConfigs: []*config.RemoteWriteConfig{&rwCfg},
+		}); err != nil {
+			return errors.Wrap(err, "applying config to remote storage")
+		}
 
-	{
-		done := make(chan struct{})
-		g.Add(func() error {
-			<-done
-			return db.Close()
-		}, func(error) {
-			close(done)
-		})
+		db, err := agent.Open(logger, reg, remoteStore, walDir, agentOpts)
```

Comment:
Could we please call this something other than `db` to avoid shadowing the outer variable? This is confusing and I had to double check this place :joy: I suggest `agentDB` or something like that, and to rename the outer `db` variable into `tsdbDB`.